Add support for Docker Registry authentication to Build

mfojtik opened the pull request:
(this is heavy WIP atm, any feedback appreciated)

mfojtik commented:
@soltysh @bparees FYI, still need heavy testing, but I would appreciate early feedback about naming/etc ;-)

mfojtik commented on pkg/build/api/types.go:
```diff
@@ -185,6 +185,11 @@ type CustomBuildStrategy struct {
+	// DockerRegistryAuthRef is used to provide Docker Registry authentication
+	// secret that is used for communication with authentication enabled Docker
+	// Registry. This have to reference existing Secret object.
+	DockerRegistrySecretRef *kapi.ObjectReference `json:"dockerRegistrySecretRef,omitempty"`
```
@bparees @soltysh I'm taking suggestion for this, I don't like this name

soltysh replied:
I'd go with just `SecretRef`

bparees replied:
Even just Secret is fine, it doesn't have to contain "Ref" in the name, we don't do that in other places that we use ObjectReference.

I'd go w/ RegistrySecret

Is this field going to exist in all 3 strategies?  if so i'm wondering if it should be factored out to a higher level.

bparees replied:
I see it does, so that question stands.  i kind of like keeping it at the same level as the image it's associated with, but i hate the fact that we have the same field 3 times.

mfojtik replied:
@bparees @soltysh created issue for that, I want to do that as follow up with other properties.

and I like RegistrySecret

mfojtik commented:
@smarterclayton PTAL

(this is pending tests which depend on having private registry with credentials support, which is currently WIP)

ncdc commented on pkg/build/api/types.go:
```diff
@@ -185,6 +185,11 @@ type CustomBuildStrategy struct {
 	// a validation error is returned.
 	Image string `json:"image"`
 
+	// DockerRegistryAuthRef is used to provide Docker Registry authentication
```
update first word

mfojtik replied:
my replace-fu failed :-( thanks

ncdc commented on pkg/build/api/types.go:
```diff
@@ -204,6 +209,11 @@ type DockerBuildStrategy struct {
 	// into the FROM line of the dockerfile.
 	Image string `json:"image,omitempty"`
+
+	// DockerRegistryAuthRef is used to provide Docker Registry authentication
```
update first word

mfojtik commented:
[test]

mfojtik commented on pkg/build/controller/strategy/util.go:
```diff
@@ -90,3 +65,49 @@ func setupBuildEnv(build *buildapi.Build, pod *kapi.Pod) error {
+	}
+
+	// TODO: Should namespace be required here?
+	namespace := secretRef.Namespace
```
@smarterclayton I'm concerned by this especially... the SecretVolumeSource require 'namespace' to be set... does that mean you have to provide namespace in the ObjectReference?

mfojtik replied:
@smarterclayton nvmd. @pmorie just explained this :)

mfojtik commented:
@liggitt PTAL

quoting @smarterclayton : build configs should probably be referencing serivce accounts

liggitt commented on pkg/build/api/validation/validation.go:
```diff
@@ -176,6 +180,17 @@ func validateStrategy(strategy *buildapi.BuildStrategy) errs.ValidationErrorList
+	if len(ref.Name) == 0 {
+		allErrs = append(allErrs, errs.NewFieldInvalid("RegistrySecret", ref.Name, "the Secret name cannot be blank"))
+	}
+	return allErrs
```
Make sure the kind is what you expect. Not sure what to check for the namespace

mfojtik replied:
If I do, then users will have to add Kind to ObjectReference... Is that
what you mean?
On Mar 23, 2015 7:34 PM, "Jordan Liggitt" <notifications@github.com> wrote:

> In pkg/build/api/validation/validation.go
> <https://github.com/openshift/origin/pull/1411#discussion_r26965430>:
>
> > @@ -176,6 +180,17 @@ func validateStrategy(strategy *buildapi.BuildStrategy) errs.ValidationErrorList
> >  	return allErrs
> >  }
> >
> > +func validateDockerSecretRef(ref *kapi.ObjectReference) errs.ValidationErrorList {
> > +	allErrs := errs.ValidationErrorList{}
> > +	if ref == nil {
> > +		return allErrs
> > +	}
> > +	if len(ref.Name) == 0 {
> > +		allErrs = append(allErrs, errs.NewFieldInvalid("RegistrySecret", ref.Name, "the Secret name cannot be blank"))
> > +	}
> > +	return allErrs
>
> Make sure the kind is what you expect. Not sure what to check for the
> namespace
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1411/files#r26965430>.
>

liggitt commented on pkg/build/controller/strategy/util.go:
```diff
@@ -90,3 +65,49 @@ func setupBuildEnv(build *buildapi.Build, pod *kapi.Pod) error {
+		return
+	}
+
+	// TODO: Should namespace be required here?
```
should a namespace other than the pod's namespace even be allowed?

mfojtik replied:
No, that should be error, I need to check that in validation.
On Mar 23, 2015 7:39 PM, "Jordan Liggitt" <notifications@github.com> wrote:

> In pkg/build/controller/strategy/util.go
> <https://github.com/openshift/origin/pull/1411#discussion_r26965923>:
>
> > @@ -90,3 +65,49 @@ func setupBuildEnv(build *buildapi.Build, pod *kapi.Pod) error {
> >  	}
> >  	return nil
> >  }
> > +
> > +// setupDockerSecrets mounts Docker Registry secrets into Pod running the build,
> > +// allowing Docker to authenticate against private registries or Docker Hub.
> > +func setupDockerSecrets(pod *kapi.Pod, secretRef *kapi.ObjectReference) {
> > +	if secretRef == nil {
> > +		return
> > +	}
> > +
> > +	// TODO: Should namespace be required here?
>
> should a namespace other than the pod's namespace even be allowed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1411/files#r26965923>.
>

liggitt commented on pkg/build/controller/strategy/util.go:
```diff
@@ -90,3 +65,49 @@ func setupBuildEnv(build *buildapi.Build, pod *kapi.Pod) error {
+
+	if len(namespace) == 0 {
+		glog.Infof("Unable to get namespace from RegistrySecret, using default")
+		namespace = kapi.NamespaceDefault
```
not sure we're even going to allow using secrets outside your namespace

mfojtik replied:
Right, it should inherit name space from Pod, which inherits it from
BuildConfig
On Mar 23, 2015 7:40 PM, "Jordan Liggitt" <notifications@github.com> wrote:

> In pkg/build/controller/strategy/util.go
> <https://github.com/openshift/origin/pull/1411#discussion_r26965974>:
>
> > +// setupDockerSecrets mounts Docker Registry secrets into Pod running the build,
> > +// allowing Docker to authenticate against private registries or Docker Hub.
> > +func setupDockerSecrets(pod *kapi.Pod, secretRef *kapi.ObjectReference) {
> > +	if secretRef == nil {
> > +		return
> > +	}
> > +
> > +	// TODO: Should namespace be required here?
> > +	namespace := secretRef.Namespace
> > +	if len(namespace) == 0 {
> > +		namespace = pod.Namespace
> > +	}
> > +
> > +	if len(namespace) == 0 {
> > +		glog.Infof("Unable to get namespace from RegistrySecret, using default")
> > +		namespace = kapi.NamespaceDefault
>
> not sure we're even going to allow using secrets outside your namespace
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1411/files#r26965974>.
>

mfojtik replied:
@liggitt I fixed this in last commit so this `if` is just a fallback (which is why it emits the message to log). The `pod.Namespace` should be set to `build.Namespace` now.

liggitt commented on pkg/build/controller/strategy/util.go:
```diff
@@ -90,3 +65,49 @@ func setupBuildEnv(build *buildapi.Build, pod *kapi.Pod) error {
+	pod.Spec.Containers[0].VolumeMounts = append(pod.Spec.Containers[0].VolumeMounts, volumeMount)
+	pod.Spec.Containers[0].Env = append(pod.Spec.Containers[0].Env, kapi.EnvVar{
+		Name:  "DOCKERCFG_PATH",
+		Value: filepath.Join(volumeMount.MountPath, secretRef.Name),
```
is this supposed to be the path to a directory or the dockercfg file itself?

mfojtik replied:
Path to file. It is read by the builder.
On Mar 23, 2015 7:41 PM, "Jordan Liggitt" <notifications@github.com> wrote:

> In pkg/build/controller/strategy/util.go
> <https://github.com/openshift/origin/pull/1411#discussion_r26966124>:
>
> > +			},
> > +		},
> > +	}
> > +	volumeMount := kapi.VolumeMount{
> > +		Name:      secretRef.Name,
> > +		MountPath: dockerSecretMountPath,
> > +		ReadOnly:  true,
> > +	}
> > +
> > +	glog.V(3).Infof("Adding volume with %s to Build pod %s", secretRef.Name, pod.Name)
> > +
> > +	pod.Spec.Volumes = append(pod.Spec.Volumes, volume)
> > +	pod.Spec.Containers[0].VolumeMounts = append(pod.Spec.Containers[0].VolumeMounts, volumeMount)
> > +	pod.Spec.Containers[0].Env = append(pod.Spec.Containers[0].Env, kapi.EnvVar{
> > +		Name:  "DOCKERCFG_PATH",
> > +		Value: filepath.Join(volumeMount.MountPath, secretRef.Name),
>
> is this supposed to be the path to a directory or the dockercfg file
> itself?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1411/files#r26966124>.
>

smarterclayton commented:
The build pods have to run as a service account (which will be set by some process, and potentially overriden by a user).  The secret used to download a Git repo or url should be one of the ones named by the account.  I think it may be ok to let users set a single named secret for the build and again one for the output.


> On Mar 23, 2015, at 12:57 PM, Michal Fojtik <notifications@github.com> wrote:
> 
> @liggitt PTAL
> 
> quoting @smarterclayton : build configs should probably be referencing serivce accounts
> 
> —
> Reply to this email directly or view it on GitHub.
>

mfojtik commented:
@smarterclayton my last commit copies the namespace from the Build itself to Build pod. Does it cover the service accounts or can you or @liggitt elaborate on what does it mean 'service account' ? Is the current mechanism to inject the secrets to build wrong?

mfojtik commented:
@smarterclayton @bparees I made some changes we were discussing yesterday, so I moved the RegistrySecret to Output and rename it to PushSecretName. This is no longer ObjectReference as the Secret can't be reference cross-namespace. 

I updated all builders to recognize two env vars, PUSH_DOCKERCFG_PATH and PULL_DOCKERCFG_PATH (one for pushing and second for pulling). At this moment the one for pushing is also used for pulling images (with big TODO to change this once we get Service Accounts)

PTAL. This is now mergeable, I tested this using my Hub credentials and the output image is pushed to Hub properly.

One problem I had was with the Secret name itself. For now, Secret.Name and the Data[Name] = "base64(dockercfg)" 'Name' needs to match. I use 'Secret.Name' as the filename for the dockercfg.

soltysh commented:
LGTM.

bparees commented on pkg/build/api/types.go:
```diff
@@ -231,6 +231,11 @@ type BuildOutput struct {
+	// pushSecretName references a name of Secret that would be used for setting
+	// up the authentication for executing the docker push to authentication
+	// enable Docker Registry (or Docker Hub).
+	PushSecretName string `json:"pushSecretName,omitempty"`
```
remove "Name"?

mfojtik replied:
@bparees thought that it would be better to be specific on what exactly we do require here ('name' of the secret)

mfojtik replied:
convinced to just PushSecret via IRC ;-)

bparees commented on pkg/build/builder/cmd/dockercfg/cfg.go:
```diff
@@ -34,14 +39,20 @@ func (h *Helper) InstallFlags(flags *pflag.FlagSet) {
 	}
 	cfg, err := readDockercfg(dockercfgPath)
 	if err != nil {
+		glog.Infof("Error reading %s: ", dockercfgPath, err)
```
info?  should be error.

some of these other logs should probably be at lower debug levels

mfojtik replied:
@bparees does it make any difference? we always start the STI builder with --loglevel=5

soltysh replied:
We should then show all the output with one level or allow user to change that. Whereas the latter I don't like personally.

mfojtik replied:
or make it inherit the log-level from the origin server itself

mfojtik replied:
@soltysh https://github.com/openshift/origin/issues/1429

mfojtik commented:
@bparees comments addressed, squashed :-)

smarterclayton commented on pkg/build/api/types.go:
```diff
@@ -231,6 +231,11 @@ type BuildOutput struct {
+	// pushSecret is the name of a Secret that would be used for setting
+	// up the authentication for executing the Docker push to authentication
+	// enabled Docker Registry (or Docker Hub).
+	PushSecret string `json:"pushSecret,omitempty"`
```
Needs to be consistent with Paul's latest, which is "SecretName".  You should not merge this until that discussion settles upstream.

mfojtik replied:
@bparees convinced me that we don't need Name as we also not using Ref
suffix.
On Mar 24, 2015 6:09 PM, "Clayton Coleman" <notifications@github.com> wrote:

> In pkg/build/api/types.go
> <https://github.com/openshift/origin/pull/1411#discussion_r27050027>:
>
> > @@ -231,6 +231,11 @@ type BuildOutput struct {
> >  	// a Docker image repository to push to. Failure to find the To will result in a build error.
> >  	To *kapi.ObjectReference `json:"to,omitempty"`
> >
> > +	// pushSecret is the name of a Secret that would be used for setting
> > +	// up the authentication for executing the Docker push to authentication
> > +	// enabled Docker Registry (or Docker Hub).
> > +	PushSecret string `json:"pushSecret,omitempty"`
>
> Needs to be consistent with Paul's latest, which is "SecretName". You
> should not merge this until that discussion settles upstream.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/1411/files#r27050027>.
>

smarterclayton commented on pkg/build/controller/strategy/util.go:
```diff
@@ -90,3 +65,45 @@ func setupBuildEnv(build *buildapi.Build, pod *kapi.Pod) error {
+	//       in order to provide the full path to dockercfg. If it does not, then
+	//       the builder fail to find the dockercfg
+	pod.Spec.Containers[0].Env = append(pod.Spec.Containers[0].Env, []kapi.EnvVar{
+		{Name: "PULL_DOCKERCFG_PATH", Value: filepath.Join(volumeMount.MountPath, pushSecret)},
```
Why not just define a convention?  I don't see why the env vars help.

bparees replied:
1) more discoverable
2) easier to change the path if we decide to do so later w/o breaking anything/anyone (eg custom builder images)

mfojtik replied:
+1 to @bparees I think documenting this properly would be sufficient for now.

smarterclayton commented on pkg/build/controller/strategy/util.go:
```diff
@@ -90,3 +65,45 @@ func setupBuildEnv(build *buildapi.Build, pod *kapi.Pod) error {
+	//       be delivered by Service Account in future.
+	// TODO: The Secret.Name must match with one if the Secret.Data[] keys in
+	//       in order to provide the full path to dockercfg. If it does not, then
+	//       the builder fail to find the dockercfg
```
Add this to the API documentation of the Secret field so that it's clear to end users (you have to have a "dockercfg" defined in your push secret).  Also, how will you handle multiple secrets with the same filenames in them?  You probably want to mount the push secret at /var/run/secrets/push (and require a dockercfg name in the secret)

mfojtik replied:
@smarterclayton thanks, I changed the MountPath to "/var/run/secrets/push"

bparees commented on pkg/build/builder/docker.go:
```diff
@@ -69,6 +72,20 @@ func (d *DockerBuilder) Build() error {
 	tag := d.build.Parameters.Output.DockerImageReference
 	defer removeImage(d.dockerClient, tag)
 	if len(d.build.Parameters.Output.DockerImageReference) != 0 {
+		// TODO: This should use ImageRepository ObjectReference in future
```
I don't think you need this... just like we resolve From to an ImageReference before creating the Build object, we should always resolve To to an ImageReference before creating the Build object.  That will ensure you can rerun the build w/ the same inputs+outputs even if the ImageRepo referred to by To or From has changed.

it should be invalid to specify To or From in the Build object itself (unfortunately our validation logic is shared right now so we need to refactor to make that happen.  See my comments here: https://github.com/openshift/origin/pull/1406#issuecomment-84725616

bparees replied:
also if it is a thing we wanted to do, then you should have the same TODO for sti.go

mfojtik replied:
@bparees will remove that TODO, my understanding was that we will refactor this to support referecences, but agree I don't need this :-)

bparees commented on pkg/build/controller/strategy/util.go:
```diff
@@ -90,3 +65,45 @@ func setupBuildEnv(build *buildapi.Build, pod *kapi.Pod) error {
+
+	// TODO: The push && pull secrets are now the same, but the pull secret should
+	//       be delivered by Service Account in future.
+	// TODO: The Secret.Name must match with one if the Secret.Data[] keys in
```
s/if/of/

bparees commented:
lgtm for handling push.

mfojtik commented:
@bparees @smarterclayton comments addressed, do you want me still wait for @pmorie name vs. not name ?

bparees commented:
@mfojtik yeah might as well have the 5 minute discussion rather than having to do another PR.

mfojtik commented:
@bparees @pmorie @smarterclayton updated, all comments addressed, so I think this is ready for push and follow up work (documentation && testing)

bparees commented:
lgtm, +1 on needs doc + testing :)

mfojtik commented:
[merge]
